Audio sync receive only - EPS8266

netmindz opened the pull request:
Option to have ESP8266 listen for AudioReactive data from an ESP32

netmindz committed:
Attempt at creating audio sync listen EPS8266

netmindz committed:
Attempt at creating audio sync listen EPS8266

netmindz commented:
Issue with soundAgc , autoResetPeak maxVol, binNum & FFT_MajPeakSmth

netmindz committed:
Attempt at creating audio sync listen EPS8266

netmindz committed:
Attempt at creating audio sync listen EPS8266

netmindz committed:
Fix compile errors for soundAgc and limitSoundDynamics - but not ideal, should only be on the sender

netmindz committed:
Fix compile errors for packet type

netmindz committed:
soundAgc should be none by default

netmindz committed:
add comment about limitSoundDynamics

netmindz committed:
Fix compile errors for um_data not present on receive node

netmindz commented on usermods/audioreactive/audio_reactive.h:
```diff
@@ -1524,6 +1538,7 @@ class AudioReactive : public Usermod {
         um_data->u_type[4] = UMT_FLOAT;
         um_data->u_data[5] = &my_magnitude;    // used (New)
         um_data->u_type[5] = UMT_FLOAT;
+#ifdef ARDUINO_ARCH_ESP32
```
Looks like the following values should really be in the UDP data

softhack007 replied:
Yes this was a remainder of the old WLED-SR branch for 8266, where fftResult[] was not computed on the small 8266. No reason any more to exclude that, as its delivered in the UDP package.

softhack007 replied:
Sorry, misunderstood the question 😅 
No maxVol and binNum need to exist as local vars, but they are not transmitted with UDP.

netmindz replied:
How do we fix then? Can we without new version of the protocol?

netmindz commented on usermods/audioreactive/audio_reactive.h:
```diff
@@ -66,6 +68,36 @@
+static unsigned long timeOfPeak = 0; // time of last sample peak detection.
+static uint8_t fftResult[NUM_GEQ_CHANNELS]= {0};// Our calculated freq. channel result table to be used by effects
+
+// TODO: probably best not used by receive nodes
+static uint8_t soundAgc = 0;                  // Automagic gain control: 0 - none, 1 - normal, 2 - vivid, 3 - lazy (config value)
```
AGC is only in the node doing the processing, so address the usage of this value on the receive node

softhack007 replied:
Agreed

netmindz replied:
Have you looked at the usages of this variable to see why I've flagged this as an issue?

softhack007 replied:
solved in latest commit.

netmindz commented on usermods/audioreactive/audio_reactive.h:
```diff
@@ -66,6 +68,36 @@
+// TODO: probably best not used by receive nodes
+static uint8_t soundAgc = 0;                  // Automagic gain control: 0 - none, 1 - normal, 2 - vivid, 3 - lazy (config value)
+
+// TODO: probably best not used by receive nodes
+// user settable parameters for limitSoundDynamics()
+static bool limiterOn = true;                 // bool: enable / disable dynamics limiter
+static uint16_t attackTime = 50;              // int: attack time in milliseconds. Default 0.08sec
+static uint16_t decayTime = 300;              // int: decay time in milliseconds.  New default 300ms. Old default was 1.40sec
```
Feels like it should be on the processing node, if not then need to update the settings page to allow these values to be configured

netmindz replied:
@softhack007 ?

softhack007 replied:
Hi 👋 ,
 you need this function on the receiver side, too. Its smoothing out the delivery delays between receiving UDP packages. Without this, animations will look wery bad (ever seen the "thriller" video from MJ?).

netmindz replied:
Ok, will keep on the recieving side then. I'll expose the config values then to the settings page

netmindz replied:
Added to config

netmindz replied:
![Screenshot from 2023-08-13 16-51-33](https://github.com/MoonModules/WLED/assets/442066/589a5a56-0fbc-4cfa-ba5c-c7b30edf7f35)

softhack007 commented:
> Issue with soundAgc , autoResetPeak maxVol, binNum & FFT_MajPeakSmth

* soundAgc: can be kept as "0", because AGC is done by the sender.
* FFT_MajPeakSmth: not transmitted yet, but you can use FFT_MajorPeak as replacement
* soundPressure: use volumeSmth as replacement
* agcSensitivity: can be replaced by "1" (not transmitted by udp)

softhack007 commented:
I can provide more feedback in the next days

softhack007 commented on usermods/audioreactive/audio_reactive.h:
```diff
@@ -1,11 +1,13 @@
+
 #include <driver/i2s.h>
 #include <driver/adc.h>
+#include <arduinoFFT.h>
```
Please keep this one lower in the code, because you have to set some #defines before including ArduinoFFT.h.

You're right it's not needed for receive-only on 8266.

netmindz replied:
reverted

softhack007 commented on usermods/audioreactive/audio_reactive.h:
```diff
@@ -303,7 +314,6 @@ static float windowWeighingFactors[samplesFFT] = {0.0f};
   // around 50% slower on -S2
 // lib_deps += https://github.com/blazoncek/arduinoFFT.git
 #endif
-#include <arduinoFFT.h>
```
It has to be exactly here, otherwise won't  work properly on esp32

softhack007 replied:
I know that some coding styles (like MISRA) strongly suggest to put all #includes at the beginning of the file. However its necessary here to do it differently.

netmindz replied:
reverted

softhack007 commented on usermods/audioreactive/audio_reactive.h:
```diff
@@ -1832,13 +1853,14 @@ class AudioReactive : public Usermod {
       }
-
+#endif
     }
```
The loop() is a bit tricky, you need to be very careful when slicing it up for UDP receive only.

I'll make a proposal for the reduced usermod::loop(), next few days.

netmindz committed:
Revert placement of arduinoFFT.h include as per comments on PR:60

netmindz committed:
Expose dynamics - limiter, rise and fall as calc runs on receive too

netmindz commented:
Think that's all the changes bar the autoResetPeak one

netmindz committed:
Fix maxVol and binNum

netmindz committed:
Fix maxVol and binNum

blazoncek commented:
I know you all hate me for my "strict selection criteria" (or plain old bluntness) but I think this PR should be done upstream.
Everyone benefits that way.

Please.

softhack007 committed:
minor fix for 8266

small typo, resulting in "A0" listed twice in the pin info.

softhack007 committed:
8266 hotfix: solving connectivity problems

* it seems that `WiFiUDP.flsuh()` does something completely different from 8266, and  its actually causing WLED to stall on UI calls. So not usable on 8266.
* fixing a few compiler warnings about "comparing signed and unsigned"

NB: its a hotfix, we have to check if there are other problems on 8266. Also its definitely ugly, but it helps as a band aid.

softhack007 committed:
agcSensitivity (to align um_data with esp32)

all elements of um_data must be filled, otherwise some effects will crash.
This adds a dummy " agcSensitivity" value that stays at 128.

softhack007 committed:
some improvements

* remove some internal variables (soundAgc, multAgc, ....)
* prototype for autoResetPeak(void)
* changed `#ifndef ESP8266` to `#ifdef ARDUINO_ARCH_ESP32`
* replacement code for computing max sample
* Info Page: "audio source" added (idle/receiving/not connected/Off)

softhack007 commented:
special cases that should be tested:
* Peak detector (samplePeak) - visible in effects, and properly reset (i.e. not stuck-at 1)
* receiving from a "v1" (SR 0.13.4) sender: AGC on, AGC off.

softhack007 committed:
8266: minor buildenv updates

* -Wno-register not needed any more (solved in FastLED 3.6.0)
* ARDUINOJSON_DEBUG added to debug_flags
* use WLEDMM_PROTECT_SERVICE on 8266, too

softhack007 committed:
show kernel debug messages in debug builds

softhack007 committed:
basic debug support on 8266

softhack007 committed:
bugfix

seems I was too fast with removing _P from strncmp_P

netmindz commented:
> I know you all hate me for my "strict selection criteria" (or plain old bluntness) but I think this PR should be done upstream. Everyone benefits that way.
> 
> Please.

Not easily possible due to significant divergence of the AR module, something I was working to reduce by at least syncing the extra drivers back upstream. The downstream change to i2c broke mm so had to revert that till I have time to see what went wrong there

netmindz commented:
> > * Peak detector (samplePeak) - visible in effects, and properly reset (i.e. not s
> > * receiving from a "v1" (SR 0.13.4) sender: AGC on, AGC off.
> 
> still to be tested.

Do you have suitable setup to test this @dosipod ?

dosipod commented:
> > > * receiving from a "v1" (SR 0.13.4) sender: AGC on, AGC off.
> > 
> > 
> > still to be tested.
> 
> Do you have suitable setup to test this @dosipod ?

@netmindz  sure we are already testing that on two units for last hour or so and it is fine , the issues we faced before was not related to audio sync  and was fixed by @softhack007
